File: livekit-examples/node-agents-examples/sip/make-a-call/src/make-call.ts
import {
  AgentDispatchClient,
  RoomServiceClient,
  SipClient,
} from 'livekit-server-sdk';

import {agentName} from './agent.js';

const {
  LIVEKIT_API_KEY = '',
  LIVEKIT_API_SECRET = '',
  LIVEKIT_URL = '',
  SIP_OUTBOUND_TRUNK_ID = '',
} = process.env;

/**
 * Formats a phone number to E.164 format for SIP compatibility
 * @param number - Raw phone number input
 * @returns Formatted phone number in E.164 format
 * @throws Error if the phone number is invalid
 */
function formatPhoneNumber(number = ''): string {
  const digits = number.replace(/\D/g, '');
  if (digits.length !== 10) {
    throw new Error(
      'Invalid phone number format. Please enter a 10-digit US phone number.',
    );
  }
  return `+1${digits}`;
}

/**
 * Makes a call to a specified phone number using LiveKit and SIP
 * @param phoneNumber - The phone number to dial
 * @throws Error if the phone number is invalid or if there is an error creating the SIP participant
 */
async function makeCall(phoneNumber: string) {
  const formattedPhoneNumber = formatPhoneNumber(phoneNumber);
  const agentDispatchClient = new AgentDispatchClient(
    LIVEKIT_URL,
    LIVEKIT_API_KEY,
    LIVEKIT_API_SECRET,
  );
  const roomServiceClient = new RoomServiceClient(
    LIVEKIT_URL,
    LIVEKIT_API_KEY,
    LIVEKIT_API_SECRET,
  );
  const sipClient = new SipClient(
    LIVEKIT_URL,
    LIVEKIT_API_KEY,
    LIVEKIT_API_SECRET,
  );

  const roomName = `room_${Math.floor(Math.random() * 10_000)}`;
  const participantIdentity = `caller_${formattedPhoneNumber}}`;

  await roomServiceClient.createRoom({name: roomName, emptyTimeout: 60});

  // Create agent dispatch
  console.log(`Creating dispatch for agent ${agentName} in room ${roomName}`);
  const dispatch = await agentDispatchClient.createDispatch(
    roomName,
    agentName,
    {
      metadata: formattedPhoneNumber,
    },
  );
  console.log(`Created dispatch: ${dispatch.id}`);

  // Create SIP participant to make the call
  if (!SIP_OUTBOUND_TRUNK_ID.startsWith('ST_')) {
    throw new Error('SIP_OUTBOUND_TRUNK_ID is not set or invalid');
  }

  console.log(`Dialing ${formattedPhoneNumber} to room ${roomName}`);

  try {
    // Create SIP participant to initiate the call
    const sipParticipant = await sipClient.createSipParticipant(
      SIP_OUTBOUND_TRUNK_ID,
      formattedPhoneNumber,
      roomName,
      {participantIdentity},
    );
    console.log(`Created SIP participant: ${sipParticipant.participantId}`);
  } catch (e) {
    console.error(
      `Error creating SIP participant: ${e instanceof Error ? e.message : e}`,
    );
  }
}

// Get phone number from command line arguments
const args = process.argv.slice(2);
const phoneNumber = args[0] ?? '';
console.log(`Dialing ${phoneNumber}`);
await makeCall(phoneNumber);

File: livekit-examples/python-agents-examples/docs/examples/make_call/make_call.py
"""
---
title: Outbound Calling Script
category: telephony
tags: [telephony, sip, outbound-calls]
difficulty: beginner
description: Script that makes outbound calls via LiveKit Telephony using the LiveKit API
demonstrates:
  - Creating agent dispatches via the API
  - Dialing outbound numbers via SIP trunks
  - LiveKit API client usage
---
"""

import asyncio
import os
import logging
from dotenv import load_dotenv
from livekit import api

load_dotenv()

logger = logging.getLogger("make-call")
logger.setLevel(logging.INFO)

room_name = "my-room"
agent_name = "test-agent"
outbound_trunk_id = os.getenv("SIP_OUTBOUND_TRUNK_ID")

async def make_call(phone_number):
    """Create a dispatch and add a SIP participant to call the phone number"""
    lkapi = api.LiveKitAPI()

    logger.info(f"Creating dispatch for agent {agent_name} in room {room_name}")
    dispatch = await lkapi.agent_dispatch.create_dispatch(
        api.CreateAgentDispatchRequest(
            agent_name=agent_name, room=room_name, metadata=phone_number
        )
    )
    logger.info(f"Created dispatch: {dispatch}")

    if not outbound_trunk_id or not outbound_trunk_id.startswith("ST_"):
        logger.error("SIP_OUTBOUND_TRUNK_ID is not set or invalid")
        return

    logger.info(f"Dialing {phone_number} to room {room_name}")

    try:
        sip_participant = await lkapi.sip.create_sip_participant(
            api.CreateSIPParticipantRequest(
                room_name=room_name,
                sip_trunk_id=outbound_trunk_id,
                sip_call_to=phone_number,
                participant_identity="phone_user",
            )
        )
        logger.info(f"Created SIP participant: {sip_participant}")
    except Exception as e:
        logger.error(f"Error creating SIP participant: {e}")

    await lkapi.aclose()

async def main():
    phone_number = "+1231231231"
    await make_call(phone_number)

if __name__ == "__main__":
    asyncio.run(main())
